# **Mengukur Jarak Data**

Mengukuran jarak data (Similarity dan Dissimilarity) adalah proses untuk menentukan seberapa mirip atau berbeda dua data atau lebih. Jarak ini dapat digunakan dalam berbagai algoritma pembelajaran mesin, seperti clustering, klasifikasi, dan regresi. Dalam mengukur jarak data terdapat beberapa tipe atribut seperti:

- Numerik

- Binary

- Categorical

- Ordinal

## **Mengukur Jarak Data (Tipe Numerik)**

Salah satu tantangan dalam era ini adalah database yang memiliki banyak tipe data. Mengukur jarak adalah komponen utama dalam algoritma clustering berbasis jarak. Algoritma partitioning seperti K-Means, K-Medoids, Fuzzy C-Means, dan Rough Clustering sangat bergantung pada perhitungan jarak untuk melakukan pengelompokkan data.

Sebelum menjelaskan tentang beberapa macam ukuran jarak, kita mendefinisikan terlebih dahulu yaitu $v_1$ dan $v_2$ menyatakan dua vektor:
$$
v_1 = (x_1, x_2, \ldots, x_n)
$$
$$
v_2 = (y_1, y_2, \ldots, y_n)
$$di mana $x_i$ dan $y_i$ disebut atribut. Ada beberapa ukuran similaritas atau ukuran jarak, di antaranya:

# **1. Minkowski Distance**
Minkowski adalah ukuran jarak umum yang mencakup Euclidean dan Manhattan sebagai kasus khusus.

$$
d(x,y) = \left( \sum_{i=1}^{n} |x_i - y_i|^m \right)^{\frac{1}{m}}, \quad m \ge 1
$$

- $m=1$ → Manhattan  
- $m=2$ → Euclidean  


# **2. Manhattan Distance**
Menghitung jarak sebagai jumlah selisih absolut tiap atribut.

$$
d(x,y) = \sum_{i=1}^{n} |x_i - y_i|
$$

# **3. Euclidean Distance**  
Jarak garis lurus antar dua titik, paling sering digunakan.

$$
d(x,y) = \sqrt{\sum_{i=1}^{n} (x_i - y_i)^2}
$$

Kelemahan: sensitif terhadap skala besar dan outlier.

# **4. Average Distance**  
Versi rata-rata dari Euclidean untuk mengurangi pengaruh jumlah dimensi.

$$
d(x,y) = \sqrt{\frac{1}{n} \sum_{i=1}^{n} (x_i - y_i)^2}
$$

# **5. Weighted Euclidean Distance**  
Memberikan bobot pada setiap atribut.

$$
d(x,y) = \sqrt{\sum_{i=1}^{n} w_i (x_i - y_i)^2}
$$


# **6. Chord Distance**  
Modifikasi Euclidean berbasis normalisasi panjang vektor.

$$
d(x,y) = \sqrt{2 - 2 \frac{\sum x_i y_i}{\|x\| \|y\|}}
$$

# **7. Mahalanobis Distance**  
Memperhitungkan korelasi antar atribut (lebih akurat untuk data kompleks).

$$
d(x,y) = \sqrt{(x - y)^T S^{-1} (x - y)}
$$


# **8. Cosine Similarity**  
Mengukur kemiripan arah (sering dipakai pada teks/dokumen).

$$
\cos(x,y) = \frac{\sum x_i y_i}{\|x\| \|y\|}
$$


# **9. Pearson Correlation**  
Mengukur hubungan linear antar dua data.

$$
r = \frac{\sum (x_i - \mu_x)(y_i - \mu_y)}{\sqrt{\sum (x_i - \mu_x)^2} \sqrt{\sum (y_i - \mu_y)^2}}
$$

Kelemahan: sensitif terhadap outlier.

##**Jarak Atribut Binary**

# **Biner Simetris**
Biner simetris adalah atribut biner di mana kedua nilai (0 dan 1) sama pentingnya. Nilai 1 tidak lebih penting dari 0, begitupun sebaliknya. Rumus menghitung jarak data dengan atribut biner simetris sebagai berikut:



$$
d(x,y) = \sum_{i=1}^{n} |x_i - y_i|
$$

Dimana:
- $x$ merupakan data pertama  
- $y$ merupakan data kedua  
- $x_i$ merupakan fitur ke-$i$ dari data pertama  
- $y_i$ merupakan fitur ke-$i$ dari data kedua  

# **Mengukur Jarak Tipe Categorical (Nominal)**
Digunakan untuk data dengan atribut kategori (misalnya: warna, jenis, dll).

# **1. Overlay Metric (OM)**
Menghitung jarak dengan cara membandingkan setiap atribut.

$$
d(x,y) = \sum_{i=1}^{n} \delta(a_i(x), a_i(y))
$$

- $\delta = 0$ jika sama  
- $\delta = 1$ jika berbeda  

Intinya: hitung berapa banyak atribut yang berbeda.

# **2. Value Difference Metric (VDM)**  
Mengukur jarak berdasarkan perbedaan probabilitas kelas dari tiap nilai atribut.

$$
d(x,y) = \sum_{i=1}^{n} \sum_{c=1}^{C} |P(c|a_i(x)) - P(c|a_i(y))|
$$

Intinya: dua nilai dianggap mirip jika sering muncul pada kelas yang sama.


#**3. Short and Fukunaga Metric (SFM)**  
Mengukur jarak berdasarkan probabilitas keseluruhan data.

$$
d(x,y) = \sum_{c=1}^{C} |P(c|x) - P(c|y)|
$$

Probabilitas biasanya dihitung dengan Naive Bayes.

#**4. Minimum Risk Metric (MRM)**  
Mengukur jarak berdasarkan risiko kesalahan klasifikasi.

$$
d(x,y) = \sum_{c=1}^{C} P(c|x)(1 - P(c|y))
$$

Intinya: semakin kecil risiko salah klasifikasi, semakin mirip datanya.

#**Mengukur Jarak Tipe Ordinal**  

Atribut ordinal adalah data yang memiliki urutan (ranking), misalnya: kecil < sedang < besar.  
Namun, jarak antar nilai tidak diketahui pasti.


Langkah Menghitung Jarak:

#**1. Ubah nilai kategori menjadi peringkat**  
Misalnya:
kecil = 1, sedang = 2, besar = 3  

$$
r_{if} \in \{1, 2, ..., M_f\}
$$


#**2. Normalisasi peringkat ke rentang [0,1]**  
Agar semua atribut memiliki skala yang sama:

$$
z_{if} = \frac{r_{if} - 1}{M_f - 1}
$$


#**3. Hitung jarak seperti data numerik**  
Setelah menjadi $z_{if}$, gunakan jarak seperti Euclidean atau Manhattan.


Intinya:  
- Data ordinal → ubah ke ranking  
- Ranking → dinormalisasi  
- Setelah itu → dihitung seperti data numerik

# **Menghitung Jarak Tipe Campuran**
Digunakan untuk data yang memiliki berbagai tipe atribut:
numerik, nominal, biner, dan ordinal.

#**Rumus Umum:**

$$
d(i,j) = \frac{\sum_{f=1}^{p} \delta_{ij}^{(f)} \, d_{ij}^{(f)}}{\sum_{f=1}^{p} \delta_{ij}^{(f)}}
$$


#**Keterangan:**
- $p$ = jumlah atribut  
- $\delta_{ij}^{(f)} = 0$ jika data hilang / tidak valid  
- $\delta_{ij}^{(f)} = 1$ jika data valid  
- $d_{ij}^{(f)}$ = jarak pada atribut ke-$f$


#**Cara Hitung Tiap Tipe Atribut:**

#1. Numerik  
$$
d_{ij}^{(f)} = \frac{|x_{if} - x_{jf}|}{\max(x_f) - \min(x_f)}
$$
Selisih dinormalisasi (agar skala sama)



#2. Nominal / Biner  
$$
d_{ij}^{(f)} =
\begin{cases}
0, & \text{jika sama} \\
1, & \text{jika berbeda}
\end{cases}
$$

Sama = 0, beda = 1



#3. Ordinal  
Ubah ke ranking lalu normalisasi:

$$
z_{if} = \frac{r_{if} - 1}{M_f - 1}
$$

Setelah itu diperlakukan seperti numerik



Intinya:
- Semua tipe data diubah ke skala yang sama (0–1)
- Hitung jarak tiap atribut
- Ambil rata-rata → jadi jarak total

Ini sering disebut juga sebagai Gower Distance

#**Implementasi Menghitung Jarak**
Menghitung jarak data numerik dari salah satu sampel data dibawah ini

# **Dataset Iris (Numerik)**

In [14]:
import pandas as pd
import numpy as np

from google.colab import files
uploaded = files.upload()

df = pd.read_csv("IRIS.csv")
df.head()

Saving IRIS.csv to IRIS (1).csv


,sepal_length,sepal_width,petal_length,petal_width,species
0,5.1,3.5,1.4,0.2,Iris-setosa
1,4.9,3.0,1.4,0.2,Iris-setosa
2,4.7,3.2,1.3,0.2,Iris-setosa
3,4.6,3.1,1.5,0.2,Iris-setosa
4,5.0,3.6,1.4,0.2,Iris-setosa


Tabel Tipe Data dan Penjelasan Kolom Dataset Iris

| Nama Kolom     | Tipe Data (Pandas) | Jenis Atribut (Statistik) | Penjelasan Fungsi Kolom                                                                                             |
| -------------- | ------------------ | ------------------------- | ------------------------------------------------------------------------------------------------------------------- |
| `sepal_length` | float64            | Numerik (Kontinu)         | Ukuran panjang kelopak bunga (sepal) dalam satuan centimeter.                                                       |
| `sepal_width`  | float64            | Numerik (Kontinu)         | Ukuran lebar kelopak bunga (sepal) dalam satuan centimeter.                                                         |
| `petal_length` | float64            | Numerik (Kontinu)         | Ukuran panjang mahkota bunga (petal) dalam satuan centimeter.                                                       |
| `petal_width`  | float64            | Numerik (Kontinu)         | Ukuran lebar mahkota bunga (petal) dalam satuan centimeter.                                                         |
| `species`      | object             | Kategorikal (Nominal)     | Label kelas (target) yang berisi teks nama spesies bunga Iris (`Iris-setosa`, `Iris-versicolor`, `Iris-virginica`). |


Ringkasan hasil deskripsi data:

| Kategori Informasi  | Jumlah             | Nama Elemen                                                             | Deskripsi Singkat                                                                     |
| ------------------- | ------------------ | ----------------------------------------------------------------------- | ------------------------------------------------------------------------------------- |
| **Dimensi Dataset** | 150 Baris, 5 Kolom | `sepal_length`, `sepal_width`, `petal_length`, `petal_width`, `species` | Total observasi sampel bunga Iris beserta seluruh variabelnya.                        |
| **Atribut**         | 4 Atribut          | `sepal_length`, `sepal_width`, `petal_length`, `petal_width`            | Variabel numerik yang digunakan sebagai fitur untuk analisis atau proses klasifikasi. |
| **Label**           | 3 Kelas            | `species` (`Iris-setosa`, `Iris-versicolor`, `Iris-virginica`)          | Variabel dependen yang menjadi tujuan akhir dari klasifikasi.                         |


**Perhitungan Manual Jarak Dataset Iris**
Berdasarkan data sampel Iris (Indeks 0 dan Indeks 1):

- P1 (Indeks 0): (5.1, 3.5, 1.4, 0.2)

- P2 (Indeks 1): (4.9, 3.0, 1.4, 0.2)

**Euclidean Distance**

### Persamaan Umum

$$
d_{Euclidean}(p,q) =
\sqrt{\sum_{i=1}^{n}(p_i - q_i)^2}
$$




### Implementasi pada fitur Iris

$$
d =
\sqrt{(SL_p - SL_q)^2 +
(SW_p - SW_q)^2 +
(PL_p - PL_q)^2 +
(PW_p - PW_q)^2}
$$




### Contoh Perhitungan

$$
d =
\sqrt{(5.1-4.9)^2 + (3.5-3.0)^2 + (1.4-1.4)^2 + (0.2-0.2)^2}
$$

$$
d =
\sqrt{(0.2)^2 + (0.5)^2 + 0^2 + 0^2}
$$

$$
d =
\sqrt{0.04 + 0.25}
$$

$$
d =
\sqrt{0.29} \approx 0.5385
$$

## Manhattan Distance

### Persamaan Umum:

$$
d_{Manhattan}(p,q) =
\sum_{i=1}^{n} |p_i - q_i|
$$

### Implementasi pada fitur Iris:

$$
d =
|SL_p - SL_q|
+
|SW_p - SW_q|
+
|PL_p - PL_q|
+
|PW_p - PW_q|
$$

### Contoh Perhitungan:

$$
d =
|5.1 - 4.9|
+
|3.5 - 3.0|
+
|1.4 - 1.4|
+
|0.2 - 0.2|
$$

$$
d =
|0.2| + |0.5| + 0 + 0
$$

$$
d =
0.2 + 0.5
$$

$$
d = 0.7000
$$

## Ringkasan Hasil

| Metode Perhitungan | Hasil Jarak |
|---|---|
| **Euclidean Distance** | **0.5385** |
| **Manhattan Distance** | **0.7000** |

# **Euclidean Distance**

In [16]:
cols = ['sepal_length', 'sepal_width', 'petal_length', 'petal_width']

p1 = df.loc[0, cols]
p2 = df.loc[1, cols]

euclidean = np.sqrt(np.sum(((p1 - p2))**2))

print(f"Hasil Jarak Euclidean : {euclidean:.4f}")

Hasil Jarak Euclidean : 0.5385


# **Manhattan Distance**

In [17]:
cols = ['sepal_length', 'sepal_width', 'petal_length', 'petal_width']

p1 = df.loc[0, cols]
p2 = df.loc[1, cols]

manhattan = np.sum(np.abs(p1 - p2))

print(f"Hasil Jarak Euclidean : {manhattan:.4f}")

Hasil Jarak Euclidean : 0.7000


# **Dataset student performance (Campuran)**

In [19]:
import pandas as pd
import numpy as np
from scipy.spatial import distance
from math import sqrt

In [20]:
from google.colab import files
uploaded = files.upload()

Saving StudentsPerformance.csv to StudentsPerformance (1).csv


In [21]:
df = pd.read_csv("StudentsPerformance.csv")
df.head()

,gender,race/ethnicity,parental level of education,lunch,test preparation course,math score,reading score,writing score
0,female,group D,master's degree,free/reduced,completed,74,72,48
1,female,group D,some college,free/reduced,completed,100,97,65
2,male,group E,high school,free/reduced,none,70,100,53
3,male,group B,master's degree,standard,completed,91,46,48
4,male,group B,some high school,free/reduced,none,84,53,83


# **Tabel Tipe Data dan Penjelasan Kolom / Atribut**

### Tabel Tipe Data dan Penjelasan Kolom (StudentsPerformance)

| Nama Kolom | Tipe Data (Pandas) | Jenis Atribut (Statistik) | Penjelasan Fungsi Kolom |
|---|---|---|---|
| `gender` | object | Kategorikal (Nominal) | Menunjukkan jenis kelamin siswa (male atau female). |
| `race/ethnicity` | object | Kategorikal (Nominal) | Kelompok ras atau etnis siswa yang dibagi menjadi beberapa grup (A–E). |
| `parental level of education` | object | Ordinal | Tingkat pendidikan terakhir orang tua siswa, misalnya high school, some college, bachelor’s degree, dll. |
| `lunch` | object | Kategorikal (Nominal) | Jenis makan siang yang diterima siswa (standard atau free/reduced). |
| `test preparation course` | object | Kategorikal (Nominal) | Menunjukkan apakah siswa mengikuti kursus persiapan ujian (completed atau none). |
| `math score` | int64 | Numerik (Kontinu) | Nilai ujian matematika siswa dengan rentang nilai 0–100. |
| `reading score` | int64 | Numerik (Kontinu) | Nilai ujian membaca siswa dengan rentang nilai 0–100. |
| `writing score` | int64 | Numerik (Kontinu) | Nilai ujian menulis siswa dengan rentang nilai 0–100. |

### Kolom yang Digunakan untuk Perhitungan Jarak

| Nama Kolom | Tipe Data (Pandas) | Jenis Atribut | Keterangan |
|---|---|---|---|
| `math score` | int64 | Numerik (Kontinu) | Nilai ujian matematika siswa. Digunakan sebagai salah satu fitur untuk menghitung jarak antar data siswa. |
| `reading score` | int64 | Numerik (Kontinu) | Nilai ujian membaca siswa. Digunakan dalam perhitungan jarak untuk melihat perbedaan kemampuan membaca. |
| `writing score` | int64 | Numerik (Kontinu) | Nilai ujian menulis siswa. Digunakan sebagai fitur tambahan dalam menghitung jarak antar siswa. |

| Data        | Math Score | Reading Score | Writing Score |
| ----------- | ---------- | ------------- | ------------- |
| **Siswa P** | 72         | 72            | 74            |
| **Siswa Q** | 69         | 90            | 88            |


## Numerik

Pada perhitungan numerik ini saya menggunakan data kolom `math score`, `reading score`, dan `writing score` untuk menghitung jarak antara dua siswa (P0 dan P1) menggunakan ukuran jarak yang telah dijelaskan sebelumnya.

---

## Euclidean Distance (Normalized)

Jarak Euclidean adalah panjang garis lurus antara dua titik dalam ruang multidimensi. Rumus umumnya adalah:

$$
d_{Euclidean}(p,q) = \sqrt{\sum_{i=1}^{n} (p_i - q_i)^2}
$$

Jika diterapkan pada fitur numerik **StudentsPerformance** dengan normalisasi berbasis Range $(R_k)$:

$$
d_{Euclidean\_Norm} =
\sqrt{
\left(\frac{Math_p - Math_q}{R_{Math}}\right)^2 +
\left(\frac{Reading_p - Reading_q}{R_{Reading}}\right)^2 +
\left(\frac{Writing_p - Writing_q}{R_{Writing}}\right)^2
}
$$

---

### Parameter Range $(R_k)$:

- $R_{Math} = 100$
- $R_{Reading} = 100$
- $R_{Writing} = 100$

Range ini diperoleh dari rentang nilai ujian siswa yang berada pada skala **0–100**.

---

### Contoh Data

Misalkan diambil dua siswa:

| Data | Math Score | Reading Score | Writing Score |
|-----|------------|---------------|---------------|
| **P0** | 72 | 72 | 74 |
| **P1** | 69 | 90 | 88 |

---

### Contoh Perhitungan

$$
d =
\sqrt{
\left(\frac{72-69}{100}\right)^2 +
\left(\frac{72-90}{100}\right)^2 +
\left(\frac{74-88}{100}\right)^2
}
$$

$$
d =
\sqrt{
(0.03)^2 + (-0.18)^2 + (-0.14)^2
}
$$

$$
d =
\sqrt{0.0009 + 0.0324 + 0.0196}
$$

$$
d =
\sqrt{0.0529}
$$

$$
d \approx 0.230
$$

In [26]:
import numpy as np
import pandas as pd

cols = ['math score','reading score','writing score']

p1 = pd.Series([72,72,74], index=cols)
p2 = pd.Series([69,90,88], index=cols)

ranges = 100

euclidean = np.sqrt(np.sum(((p1 - p2) / ranges)**2))

print(f"Jarak Euclidean (Normalized) : {euclidean:.4f}")

Jarak Euclidean (Normalized) : 0.2300


## Perhitungan Manhattan Distance (Normalized)

Jika diterapkan pada fitur numerik **StudentsPerformance** dengan normalisasi berbasis Range $(R_k)$:

$$
d_{Manhattan\_Norm} = \sum_{k=1}^{n} \frac{|p_k - q_k|}{R_k}
$$

Jika diterapkan pada fitur nilai siswa:

$$
d_{Manhattan\_Norm} =
\frac{|Math_p - Math_q|}{R_{Math}} +
\frac{|Reading_p - Reading_q|}{R_{Reading}} +
\frac{|Writing_p - Writing_q|}{R_{Writing}}
$$

---

### Parameter Range $(R_k)$:

- $R_{Math} = 100$
- $R_{Reading} = 100$
- $R_{Writing} = 100$

Range ini diperoleh dari rentang nilai ujian siswa yang berada pada skala **0–100**.

---

### Substitusi nilai ke dalam rumus

Misalkan dua siswa:

| Data | Math | Reading | Writing |
|-----|------|--------|--------|
| **P0** | 72 | 72 | 74 |
| **P1** | 69 | 90 | 88 |

Substitusi:

$$
d =
\frac{|72 - 69|}{100} +
\frac{|72 - 90|}{100} +
\frac{|74 - 88|}{100}
$$

$$
d =
\frac{3}{100} +
\frac{18}{100} +
\frac{14}{100}
$$

$$
d =
0.03 + 0.18 + 0.14
$$

$$
d = 0.35
$$

In [28]:
import pandas as pd
import numpy as np

cols = ['math score', 'reading score', 'writing score']

p1 = pd.Series([72, 72, 74], index=cols)
p2 = pd.Series([69, 90, 88], index=cols)

ranges = 100

manhattan = np.sum(np.abs(p1 - p2) / ranges)

print(f"Jarak Manhattan (Normalized) : {manhattan:.4f}")

Jarak Manhattan (Normalized) : 0.3500


## Binary & Categorical

Pada perhitungan jarak untuk tipe data **binary dan categorical**, saya menggunakan data kolom `gender` (Binary) dan `lunch` (Categorical) untuk menghitung jarak antara dua siswa (P0 dan P1) menggunakan ukuran jarak yang telah dijelaskan sebelumnya.

Pada tipe data ini, tidak ada urutan atau besaran angka. Jarak hanya dihitung berdasarkan apakah nilai tersebut **sama atau berbeda**.

Jarak antara dua data $p$ dan $q$ untuk fitur kategorikal/biner didefinisikan sebagai:

$$
d(p,q) =
\begin{cases}
0, & \text{jika } p = q \\
1, & \text{jika } p \neq q
\end{cases}
$$


### Berdasarkan data sampel

- **P0:** `gender` = female, `lunch` = standard  
- **P1:** `gender` = female, `lunch` = free/reduced



# **1. Fitur Binary (`gender`)**

Perbandingan nilai untuk fitur jenis kelamin antara siswa P0 dan siswa P1:

- $p_{gender}$ = female  
- $q_{gender}$ = female  

Karena nilai keduanya **sama**, maka:

$$
d_{gender}(p,q) = 0
$$



# **2. Fitur Categorical (`lunch`)**

Perbandingan nilai untuk fitur jenis makan siang antara siswa P0 dan siswa P1:

- $p_{lunch}$ = standard  
- $q_{lunch}$ = free/reduced  

Karena nilai keduanya **berbeda**, maka:

$$
d_{lunch}(p,q) = 1
$$



# **Hasil Jarak**

Total jarak untuk fitur binary dan categorical dihitung sebagai:

$$
d = d_{gender} + d_{lunch}
$$

$$
d = 0 + 1
$$

$$
d = 1
$$

## Ordinal

Pada perhitungan jarak untuk tipe data **ordinal**, saya menggunakan data kolom `parental level of education` untuk menghitung jarak antara dua siswa (P0 dan P1) menggunakan ukuran jarak yang telah dijelaskan sebelumnya.

## Perhitungan Jarak Ordinal (Parental Level of Education)

Pada fitur `parental level of education`, terdapat beberapa tingkatan pendidikan orang tua. Kita akan mengikuti prosedur **normalisasi peringkat** agar nilainya berada dalam rentang **[0,1]**.

Urutan tingkat pendidikan dapat ditetapkan sebagai berikut:

1. some high school  
2. high school  
3. some college  
4. associate's degree  
5. bachelor's degree  
6. master's degree  


# **1. Menentukan Peringkat $(r_{if})$ dan Nilai Maksimum $(M_f)$**

Berdasarkan dataset, terdapat **6 tingkatan pendidikan**, sehingga nilai maksimum:

$$
M_f = 6
$$

Misalnya dua siswa:

- **P0:** parental level of education = bachelor's degree  
- **P1:** parental level of education = some college  

Maka peringkatnya:

- $r_{0,edu} = 5$  
- $r_{1,edu} = 3$



# **2. Normalisasi Nilai $(z_{if})$**

Menggunakan rumus:

$$
z_{if} = \frac{r_{if} - 1}{M_f - 1}
$$


# **Perhitungan Normalisasi**

Untuk siswa **P0**:

$$
z_{0} = \frac{5-1}{6-1} = \frac{4}{5} = 0.8
$$

Untuk siswa **P1**:

$$
z_{1} = \frac{3-1}{6-1} = \frac{2}{5} = 0.4
$$



# **3. Menghitung Jarak Manhattan pada Nilai $z_{if}$**

Jarak ordinal antara dua siswa P0 dan P1 dihitung sebagai selisih absolut dari nilai yang telah dinormalisasi:

$$
d_{Ordinal}(p,q) = |z_0 - z_1|
$$

Berdasarkan hasil normalisasi sebelumnya:

$$
z_0 = 0.8
$$

$$
z_1 = 0.4
$$

Sehingga:

$$
d_{Ordinal} = |0.8 - 0.4|
$$

$$
d_{Ordinal} = 0.4
$$

In [30]:
edu_rank = {
    "some high school": 1,
    "high school": 2,
    "some college": 3,
    "associate's degree": 4,
    "bachelor's degree": 5,
    "master's degree": 6
}

p1_class = edu_rank["bachelor's degree"]   # 5
p2_class = edu_rank["some college"]        # 3

M = max(edu_rank.values())

z1 = (p1_class - 1) / (M - 1)
z2 = (p2_class - 1) / (M - 1)

dist_ordinal = abs(z1 - z2)

print(f"Jarak Ordinal (Manhattan) : {dist_ordinal:.4f}")

Jarak Ordinal (Manhattan) : 0.4000


## **Gower’s Similarity Coefficient (Tipe Campuran)**

Berdasarkan rumus Gower, kita akan menjumlahkan seluruh kontribusi jarak dari setiap fitur $(d_{ij}^{(f)})$ lalu membaginya dengan jumlah fitur $(p)$.

### **1. Rekapitulasi Kontribusi Jarak $(d_{ij}^{(f)})$**

| Fitur (f) | Tipe | Nilai P0 | Nilai P1 | Perhitungan | Skor $(d_{ij}^{(f)})$ |
|---|---|---|---|---|---|
| math score | Numerik | 72 | 69 | $\frac{|72-69|}{100}$ | 0.03 |
| reading score | Numerik | 72 | 90 | $\frac{|72-90|}{100}$ | 0.18 |
| writing score | Numerik | 74 | 88 | $\frac{|74-88|}{100}$ | 0.14 |
| gender | Biner | female | female | female = female | 0.0000 |
| lunch | Nominal | standard | free/reduced | standard ≠ free/reduced | 1.0000 |
| parental level of education | Ordinal | bachelor's degree | some college | $|0.8 - 0.4|$ | 0.4000 |




## **2. Substitusi ke Rumus Gower**

Sesuai rumus:

$$
d(i,j) =
\frac{\sum_{f=1}^{p} \delta_{ij}^{(f)} d_{ij}^{(f)}}
{\sum_{f=1}^{p} \delta_{ij}^{(f)}}
$$

Karena semua data tersedia $(\delta_{ij}=1)$ dan total fitur $(p)$ adalah **6**, maka:

$$
d(0,1) =
\frac{
0.03 + 0.18 + 0.14 + 0 + 1 + 0.4
}{6}
$$

$$
d(0,1) =
\frac{1.75}{6}
$$

$$
d(0,1) \approx 0.2917
$$

In [34]:
# ambil dua data
p1 = df.loc[0]
p2 = df.loc[1]

# --- Numerik ---
dist_math = abs(p1['math score'] - p2['math score']) / 100
dist_reading = abs(p1['reading score'] - p2['reading score']) / 100
dist_writing = abs(p1['writing score'] - p2['writing score']) / 100

# --- Binary ---
dist_gender = 0 if p1['gender'] == p2['gender'] else 1

# --- Nominal ---
dist_lunch = 0 if p1['lunch'] == p2['lunch'] else 1

# --- Ordinal ---
edu_rank = {
    "some high school":1,
    "high school":2,
    "some college":3,
    "associate's degree":4,
    "bachelor's degree":5,
    "master's degree":6
}

r1 = edu_rank[p1['parental level of education']]
r2 = edu_rank[p2['parental level of education']]

M = max(edu_rank.values())

z1 = (r1 - 1) / (M - 1)
z2 = (r2 - 1) / (M - 1)

dist_edu = abs(z1 - z2)

In [36]:
import pandas as pd

p1 = pd.Series({
    'math score': 72,
    'reading score': 72,
    'writing score': 74,
    'gender': 'female',
    'lunch': 'standard',
    'parental level of education': "bachelor's degree"
})

p2 = pd.Series({
    'math score': 69,
    'reading score': 90,
    'writing score': 88,
    'gender': 'female',
    'lunch': 'free/reduced',
    'parental level of education': 'some college'
})

dist_math = abs(p1['math score'] - p2['math score']) / 100
dist_reading = abs(p1['reading score'] - p2['reading score']) / 100
dist_writing = abs(p1['writing score'] - p2['writing score']) / 100

dist_gender = 0 if p1['gender'] == p2['gender'] else 1
dist_lunch = 0 if p1['lunch'] == p2['lunch'] else 1

edu_rank = {
    "some high school": 1,
    "high school": 2,
    "some college": 3,
    "associate's degree": 4,
    "bachelor's degree": 5,
    "master's degree": 6
}

r1 = edu_rank[p1['parental level of education']]
r2 = edu_rank[p2['parental level of education']]

M = max(edu_rank.values())
z1 = (r1 - 1) / (M - 1)
z2 = (r2 - 1) / (M - 1)
dist_edu = abs(z1 - z2)

list_jarak = [dist_math, dist_reading, dist_writing, dist_gender, dist_lunch, dist_edu]
hasil_gower = sum(list_jarak) / 6

print(f"Total Jarak Gower (P0 vs P1): {hasil_gower:.4f}")

Total Jarak Gower (P0 vs P1): 0.2917


## **Kesimpulan Akhir**

Hasil akhir perhitungan **Gower Distance** antara Siswa 0 dan Siswa 1 adalah **0.2133**.

Nilai ini merepresentasikan rata-rata ketidaksamaan dari seluruh tipe atribut yang digunakan, yaitu atribut **numerik** (math score, reading score, writing score), **binary** (gender), **kategorikal** (lunch), dan **ordinal** (parental level of education).

Semakin mendekati nilai **0**, maka kedua objek semakin **serupa**, sedangkan nilai yang mendekati **1** menunjukkan perbedaan yang semakin besar.

Dalam kasus ini, skor **0.2133** menunjukkan bahwa kedua siswa memiliki **tingkat kemiripan yang cukup tinggi**, karena nilai jaraknya relatif kecil. Perbedaan yang muncul terutama dipengaruhi oleh variasi pada nilai ujian dan beberapa atribut kategorikal, sementara beberapa atribut lainnya memiliki nilai yang sama atau mendekati.